src/v/cluster/types.h

```diff
@@ -1253,6 +1255,8 @@ struct topic_properties
     // one way or another.  There is no "use the cluster default".
     bool remote_delete{storage::ntp_config::default_remote_delete};
 
+    tristate<std::chrono::milliseconds> segment_ms{std::nullopt};
```

Comment:
I think there's a few thing missing from this commit (can be broken up into several):
* `topic_properties::has_overrides` and `topic_properties::default_overrides` need to be made aware of `segment_ms`
* you'll likely have to update the compat tests in `cluster_compat.h` (see `compact_check<topic_properties>`)
* the kafka handlers for topic creation, describing, and altering configuration need to become aware of the new property

See for example https://github.com/redpanda-data/redpanda/pull/6613 which introduced the local target retention configs.